src/extension/infer.rs

```diff
@@ -663,31 +680,85 @@ impl UnificationContext {
-        for m in self.variables.clone().into_iter() {
+        // A directed graph to keep track of `Plus` constraint relationships
+        let mut relations = GraphContainer::new_directed();
+        let mut solutions: HashMap<Meta, ExtensionSet> = HashMap::new();
+
+        let variable_scope = self.search_variable_deps();
+        for m in variable_scope.into_iter() {
             if !self.solved.contains_key(&m) {
                 // Handle the case where the constraints for `m` contain a self
                 // reference, i.e. "m = Plus(E, m)", in which case the variable
                 // should be instantiated to E rather than the empty set.
-                let solution =
-                    ExtensionSet::from_iter(self.get_constraints(&m).unwrap().iter().filter_map(
-                        |c| match c {
-                            // If `m` has been merged, [`self.variables`] entry
-                            // will have already been updated to the merged
-                            // value by [`self.merge_equal_metas`] so we don't
-                            // need to worry about resolving it.
-                            Constraint::Plus(x, other_m) if m == self.resolve(*other_m) => {
-                                Some(x.clone())
-                            }
-                            _ => None,
-                        },
-                    ));
-                self.add_solution(m, solution);
+                let mut solution = ExtensionSet::new();
+                // Is the solution complete - does this variable *only* refer to itself?
+                let mut solution_complete = true;
+
+                for c in self.get_constraints(&m).unwrap().iter() {
+                    if let Constraint::Plus(r, other_m) = c {
+                        // If `m` has been merged, [`self.variables`] entry
+                        // will have already been updated to the merged
+                        // value by [`self.merge_equal_metas`] so we don't
+                        // need to worry about resolving it.
+                        let other_m = self.resolve(*other_m);
+                        solution.insert(r);
+                        if m != other_m {
+                            // This `m` depends on the solution of `other_m` so
+                            // `solution` is not enough info on its own.
+                            solution_complete = false;
+                            relations.add_edge(m, self.resolve(other_m));
+                        }
+                    }
+                }
+                solutions.insert(m, solution.clone());
+                if solution_complete {
+                    self.add_solution(m, solution);
+                }
+            }
+        }
+
+        // Strongly connected components are looping constraint dependencies.
+        // This means that each metavariable in the CC has the same solution.
+        for cc in relations.ccs() {
```

Comment:
(There are other ways, you could record the `m`s for which the first loop failed to find a solution, then tick them off here. But that'd be more expensive when debug_assert is disabled)

Reply:
We could add an assert that the number of nodes in `relations.css()` (flattened) is the same as the length of `variable_scope`?

Reply:
You'd need to filter out the number of those that were already in `self.solved` (there's an `if !self.solved.contains`)...

But the flattening idea is good - how about just that the number of nodes in flattened-`relations.css()` is equal to the number of *edges* in relations?

Reply:
Ah ok, some edges share endpoints. This works:
```suggestion
   assert_eq!(relations.node_map.len(), relations.ccs().iter().map(Vec::len).sum::<usize>());
```
Common up the two `relations.ccs` tho!

Reply:
I'm not sure that check really makes sense, since it's apparent that the size of `node_map` and the number of nodes in the graph will be the same from looking at `GraphContainer::add_or_retrieve`, and `scc`s will return a clustered version of every node in the graph

Reply:
Agreed size of node map equals number of nodes in the graph, yes the latter is what I meant.

But SCC = Strongly Connected Component, i.e. every node reaches every other? In a directed graph? Ah, are `ccs` undirectly-connected?

So A -> B -> C -> A is one SCC, with all three nodes. But A -> B -> C is NOT an SCC by that definition, but are you saying you'd get A as one `cc`, B as one `cc` and C as one `cc` ?

Reply:
Yes, indeed it does. In which case yeah, no assert/check really seems necessary, we're only really protecting against somehow getting in a muddle ourselves between the two loops.

Does make me wonder whether we could just put *everything* through the `relations` map, then. A node with no connections will trivially (and cheaply) be its own SCC. So this above:
```
if unresolved_metas,is_empty() {
   self.add_solution(m, solution.clone())
} else {
  unresolved_metas.iter().for_each(.....)
}
```
could just be
```
relations.add_or_retrieve(m);
unresolved_metas.iter().for_each(....)
```
and then all the `add_solution` calls would be in the second loop.